src/dune_rpc_impl/server.ml

```diff
@@ -146,12 +143,57 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
   let on_terminate session =
     let t = Fdecl.get t in
     t.clients <- Clients.remove_session t.clients session;
-    Long_poll.disconnect_session t.long_poll session
+    Fiber.return ()
   in
   let rpc =
     Handler.create ~on_terminate ~on_init
       ~version:Dune_rpc_private.Version.latest ()
   in
+  let () =
+    let module Diagnostic = Dune_rpc.Diagnostic in
+    let module Error = Build_config.Error in
+    let equal (s : Error.Set.t) (t : Error.Set.t) = Int.equal s.stamp t.stamp in
+    let diff ~last ~(now : Error.Set.t) =
+      match last with
+      | None ->
+        Error.Id.Map.to_list_map now.current ~f:(fun _ e ->
+            Diagnostic.Event.Add (Diagnostics.diagnostic_of_error e))
+      | Some (prev : Error.Set.t) -> (
+        match prev.last_event with
+        | Some last_event when prev.stamp + 1 = now.stamp ->
+          [ Diagnostics.diagnostic_event_of_error_event last_event ]
+        | _ ->
+          (* the slow path where we must calculate a diff between what we have
+             and the last thing we've sent to the poller *)
+          Error.Id.Map.merge prev.current now.current ~f:(fun _ prev now ->
+              match (prev, now) with
+              | None, None -> assert false
+              | Some prev, None ->
+                Some (Diagnostics.diagnostic_event_of_error_event (Remove prev))
+              | _, Some next ->
+                Some (Diagnostics.diagnostic_event_of_error_event (Add next)))
+          |> Error.Id.Map.values)
+    in
+    Handler.implement_long_poll rpc Procedures.Poll.diagnostic
+      Build_system.errors ~equal ~diff
+  in
+  let () =
+    let diff ~last:_ ~(now : Build_system.State.t) =
+      match now with
+      | Initializing -> Progress.Waiting
+      | Interrupting_current_build -> Interrupted
+      | Build_finished_with_success_and_waiting -> Success
+      | Build_finished_with_failure_and_waiting -> Failed
+      | Building now ->
+        let remaining =
+          now.number_of_rules_discovered - now.number_of_rules_executed
```

Comment:
It's a bit weird that we have some code duplication here (with `build_system.ml`).

Reply:
Do you mean with the type `Build_system.State.t`? Yes it's a bit annoying but the analogous type in RPC has stability guarantees that this type does not.